### DISCUSSION TITLE
callable object arguments - functions take the simmer env as an argument

### DISCUSSION THREAD

**Discussion by tomjemmett** [r-simmer/simmer#252] [Ideas]
If we want to use a function like `get_selected()` or `now()` we need to have access to the simmer environment, for example:

``` r
env <- simmer()

env |>
  add_resource("a", 1) |>
  add_resource("b", 1) |>
  add_generator(
    "person",
    trajectory() |>
      select(c("a", "b")) |>
      log_(\() paste("I selected", get_selected(env))) |>
      seize_selected() |>
      timeout(5) |>
      release_selected(),
    at(1, 2)
  ) |>
  run()
```

this isn't pipe friendly, so one approach is to create a function to replace add_generator, e.g.

``` r
env <- simmer() |>
  add_resource("a", 1) |>
  add_resource("b", 1) |>
  (\(.env) {
    add_generator(
      .env,
      "person",
      trajectory() |>
        select(c("a", "b")) |>
        log_(\() paste("I selected", get_selected(.env))) |>
        seize_selected() |>
        timeout(5) |>
        release_selected(),
      at(1, 2)
    )
  })() |>
  run()
```

as far as I can see, all of the functions like log_ (in this example) which can take a function for the `value` arguments have no arguments - if instead these functions were to have as their first argument the current simmer environment we could have much neater code, e.g.

``` r
env <- simmer() |>
  add_resource("a", 1) |>
  add_resource("b", 1) |>
  add_generator(
    "person",
    trajectory() |>
      select(c("a", "b")) |>
      log_(\(.env) paste("I selected", get_selected(.env))) |>
      seize_selected() |>
      timeout(5) |>
      release_selected(),
    at(1, 2)
  ) |>
  run()
```

this could cause problems for existing code, but a check to see whether the argument is a function with 0 arguments or not could solve this, e.g.

``` r
my_fn <- function(fn) {
  if (length(formals(fn)) == 0) {
    fn()
  } else {
    fn(list("hi"))
  }
}

my_fn(function() "nothing")
my_fn(function(.env) .env)
```

is this something that would be a good addition to `{simmer}`?

**Comment by Enchufa2** [comment]
I prefer explicit over implicit, but if the implementation doesn't break current functionality, this is something we could look into. I'm not sure whether this could work with things like `set_trajectory()` though.

**Reply by tomjemmett** [reply]
I agree re explicit over implicit. Though I think it would be a lot neater to always be able to work with a pipeable call :-)

I'm not sure I see this as being useful for `set_trajectory`, but here is an example I currently have:

``` r
  ... |>
  timeout(\() get_attribute(.env, "task_time")) |>
```

I have the timeout value stored in an attribute (I need to use the time elsewhere in a `trap` handler). In this case I need access to the current environment, so I have to wrap the whole generator in a function so I can add it to a pipeable call.

**Reply by tomjemmett** [reply]
(should note, I just realised this example could use `timeout_from_attribute`, but my actual code has a bit more logic)

**Reply by Enchufa2** [reply]
I'm not sure either whether this can work as you intend in your third example with some trickery in the activity function, because there's no environment. Note that your second example works because you are explicitly generating an environment where the trajectory is defined, and there's an `.env` object declared there. In your third example, the enclosing environment for the trajectory is the global environment. So how could you capture `.env`? Ideas welcome. :)

**Reply by tomjemmett** [reply]
sorry, cropped my last example, it's more like

``` r
simmer() |>
  (\(.env) {
    add_generator(
      .env,
     "x",
     trajectory() |>
      set_attribute("task_time", 5) |>
      timeout(\() get_attribute(.env, "task_time") + sample(1:3, 1)),
    at(0)
    )
  })()
 ```
 
using an immediately executed anonymous function here, but could work equally well with a named function. My proposal would lead to neater (IMHO) code:

``` r
simmer() |>
  add_generator(
    .env,
    "x",
    trajectory() |>
      set_attribute("task_time", 5) |>
      timeout(\(.env) get_attribute(.env, "task_time") + sample(1:3, 1)),
    at(0)
  )
```

This would then not need the `simmer()` environment to have been declared in the parent environment.

**Reply by Enchufa2** [reply]
Yes, I understand the proposal. What I'm saying is that I don't see how to make it work just by modifying how the activities execute the functions provided, because it's not a matter of whether the function passed has an argument or not; it's a matter of whether the correct simulation object is defined in the enclosing environment or not.

In your last example, if you don't _postpone_ the construction of the trajectory object somehow, the enclosing environment is the global environment, and thus there's no way to _see_ it in that `get_attribute` call. We need to evaluate the trajectory lazily. For instance, this seems to work:

```r
add_generator <- function(.env, name_prefix, trajectory, distribution, ...) {
  trajectory <- eval(substitute(trajectory))
  distribution <- eval(substitute(distribution))
  simmer::add_generator(.env, name_prefix, trajectory, distribution, ...)
}

env <- simmer() %>%
  add_resource("a", 1) %>%
  add_resource("b", 1) %>%
  add_generator(
    "person",
    trajectory() %>%
      select(c("a", "b")) %>%
      log_(function() paste("I selected", get_selected(.env))) %>%
      seize_selected() %>%
      timeout(5) %>%
      release_selected(),
    at(1, 2)
  ) %>%
  run()
```

Another option would be to use formula notation to define the trajectory, which would ensure the trajectory is lazy-evaluated explicitly; and it's safer, because it would maintain current behaviour. E.g.:

```r
add_generator <- function(.env, name_prefix, trajectory, distribution, ...) {
  if (inherits(trajectory, "formula"))
    trajectory <- eval(trajectory[[2]])
  if (inherits(distribution, "formula"))
    distribution <- eval(distribution[[2]])
  simmer::add_generator(.env, name_prefix, trajectory, distribution, ...)
}

env <- simmer() %>%
  add_resource("a", 1) %>%
  add_resource("b", 1) %>%
  add_generator(
    "person",
    ~ trajectory() %>%
      select(c("a", "b")) %>%
      log_(function() paste("I selected", get_selected(.env))) %>%
      seize_selected() %>%
      timeout(5) %>%
      release_selected(),
    at(1, 2)
  ) %>%
  run()
```

But all this only plays well if you inline everything in a single pipe chain. If your simulation is more complex, and requires combining several trajectory objects in several steps, it doesn't work anymore I think (didn't try though).
